File: poc-d04/src/poc_d04/gate.py
"""게이트 감사 (기획안 §3-2). 추출 결과가 원문과 규칙을 벗어났는지 결정적으로 검사한다.

G1 원문 없는 값 / 원문 위치 불일치
G2 모호 표현을 날짜로 단정
G3 별칭만으로 규격(품목) 병합
G4 원문 없는 규격 채움
G5 사람 확정 칸을 AI가 채움
"""
from __future__ import annotations

import re

from .catalog import LEXICON, VAGUE_DATE_TERMS
from .extract import SPEC_RE
from .rules import parse_date

_DATE_LIKE = re.compile(r"\d{4}-\d{2}-\d{2}|\d{1,2}월\s?\d{1,2}일|\b\d{1,2}[/.]\d{1,2}\b")
_ID = re.compile(r"P-\d{2}")
_SPEC_TOKEN = re.compile(r"\b\d{2,3}A\b|\b\d{1,2}mm\b")

GATE_TEXT = {
    "G1": "값이 없으면 비우거나 '확인 필요'. 원문 없는 값을 채우지 않는다.",
    "G2": "'다음 주' 같은 모호 표현을 날짜로 단정하지 않는다.",
    "G3": "별칭이 비슷하다는 이유만으로 다른 규격을 합치지 않는다.",
    "G4": "원문에 없는 규격을 채우지 않는다.",
    "G5": "최종 확정은 사람이 한다. AI 출력은 '제안' 열에만 둔다.",
}


def _v(code: str, field: str | None, detail: str) -> dict:
    return {"코드": code, "게이트": GATE_TEXT[code], "항목ID": field, "설명": detail}


def _in_text(text: str, value: str) -> bool:
    return value is not None and str(value) in text


def audit(extraction: dict, text: str) -> list[dict]:
    viol: list[dict] = []
    vague_in_text = [t for t in VAGUE_DATE_TERMS if t in text]
    spec_in_text = SPEC_RE.search(text) is not None

    for it in extraction["항목"]:
        f, val = it["항목ID"], it.get("원문값")
        s, e = it.get("원문시작"), it.get("원문끝")

        # G4: 규격을 원문 없이 채움 (G1보다 먼저, 별도 보고)
        if f == "spec" and val is not None and not _in_text(text, val):
            viol.append(_v("G4", f, f"규격 '{val}'이 원문에 없음"))
            continue

        # G2: 모호 표현을 날짜로 단정
        if f == "need_date":
            texts = [str(x) for x in (val, it.get("표준명후보"), it.get("메모")) if x]
            fabricated = next(
                (dm for tx in texts for dm in _DATE_LIKE.findall(tx)
                 if not _in_text(text, dm) and parse_date(dm) is not None), None)
            if fabricated:
                why = f"'{vague_in_text[0]}'" if vague_in_text else "원문"
                viol.append(_v("G2", f, f"{why}을(를) 날짜 '{fabricated}'로 단정 (원문에 없음)"))
                continue

        # G1: 원문 위치·값 추적
        if val is not None:
            if s is None or e is None:
                viol.append(_v("G1", f, f"'{val}'에 원문 위치가 없음"))
            elif text[s:e] != str(val):
                viol.append(_v("G1", f, f"원문 위치 [{s}:{e}]='{text[s:e]}' ≠ 값 '{val}'"))

        # G3: 별칭 병합
        if f == "item_id" and val is not None and not _ID.fullmatch(str(val).upper()):
            alias_cands = [x for x in LEXICON if str(val) in x.aliases or str(val) == x.name]
            cand_txt = str(it.get("표준명후보") or "")
            single = set(_ID.findall(cand_txt))
            cand_specs = {s for s in _SPEC_TOKEN.findall(cand_txt) if s not in text}
            if len(alias_cands) > 1 and not spec_in_text:
                if len(single) == 1:
                    viol.append(_v("G3", f, f"별칭 '{val}'은 {len(alias_cands)}개 품목에 해당하나 규격 없이 {single.pop()}로 확정"))
                elif len(cand_specs) == 1 and len({x.spec for x in alias_cands}) > 1:
                    viol.append(_v("G3", f, f"별칭 '{val}'의 후보에 원문에 없는 규격 '{cand_specs.pop()}'을 붙여 한 품목으로 좁힘"))

        # G5: 사람 확정 칸 선점
        if it.get("사람확정") not in (None, ""):
            viol.append(_v("G5", f, f"사람 확정 칸에 AI 값 '{it['사람확정']}'"))

    if extraction.get("사람확정") not in (None, ""):
        viol.append(_v("G5", None, "문서 수준 사람 확정 칸이 채워져 있음"))
    return viol

File: poc/src/domain/gate.ts
import { itemsById } from "./catalog.ts"
import { DATE_LIKE, parseDate, vagueTermsIn } from "./dates.ts"
import { SPEC_RE } from "./extract-rule.ts"
import type { Extraction, FieldId, GateCode, GateViolation } from "./types.ts"

export const GATE_TEXT: Record<GateCode, string> = {
  G1: "값이 없으면 비우거나 '확인 필요'. 원문 없는 값을 채우지 않는다.",
  G2: "'다음 주' 같은 모호 표현을 날짜로 단정하지 않는다.",
  G3: "별칭이 비슷하다는 이유만으로 다른 규격을 합치지 않는다.",
  G4: "원문에 없는 규격을 채우지 않는다.",
  G5: "최종 확정은 사람이 한다. AI 출력은 '제안' 열에만 둔다.",
}

const ID_TOKEN = /P-\d{2}/g
const SPEC_TOKEN = /\b\d{2,3}A\b|\b\d{1,2}mm\b/g

function violation(
  code: GateCode,
  fieldId: FieldId | null,
  detail: string,
): GateViolation {
  return { code, fieldId, detail, gate: GATE_TEXT[code] }
}

function inText(text: string, value: string | null | undefined): boolean {
  return value !== null && value !== undefined && text.includes(String(value))
}

export function auditExtraction(extraction: Extraction, text: string): GateViolation[] {
  const viol: GateViolation[] = []
  const vagueInText = vagueTermsIn(text)
  const specInText = new RegExp(SPEC_RE.source).test(text)
  const lexicon = itemsById()

  for (const item of extraction.items) {
    const fieldId = item.fieldId
    const value = item.sourceValue
    const start = item.start
    const end = item.end

    if (fieldId === "spec" && value !== null && !inText(text, value)) {
      viol.push(violation("G4", fieldId, `규격 '${value}'이 원문에 없음`))
      continue
    }

    if (fieldId === "need_date") {
      const texts = [value, item.aliasCandidate, item.memo].filter(Boolean).map(String)
      let fabricated: string | null = null
      for (const tx of texts) {
        const matches = tx.match(new RegExp(DATE_LIKE.source, "g")) ?? []
        for (const token of matches) {
          if (!inText(text, token) && parseDate(token) !== null) {
            fabricated = token
            break
          }
        }
        if (fabricated) break
      }
      if (fabricated) {
        const why = vagueInText[0] ?? "원문"
        viol.push(
          violation("G2", fieldId, `'${why}'을(를) 날짜 '${fabricated}'로 단정 (원문에 없음)`),
        )
        continue
      }
    }

    if (value !== null) {
      if (start === null || end === null) {
        viol.push(violation("G1", fieldId, `'${value}'에 원문 위치가 없음`))
      } else if (text.slice(start, end) !== String(value)) {
        viol.push(
          violation(
            "G1",
            fieldId,
            `원문 위치 [${start}:${end}]='${text.slice(start, end)}' ≠ 값 '${value}'`,
          ),
        )
      }
    }

    if (fieldId === "item_id" && value !== null && !/^P-\d{2}$/.test(value.toUpperCase())) {
      const aliasCands = Object.values(lexicon).filter(
        (entry) => entry.aliases.includes(value) || entry.name === value,
      )
      const candTxt = item.aliasCandidate ?? ""
      const single = new Set(candTxt.match(ID_TOKEN) ?? [])
      const candSpecs = [...(candTxt.match(SPEC_TOKEN) ?? [])].filter((s) => !text.includes(s))
      if (aliasCands.length > 1 && !specInText) {
        if (single.size === 1) {
          const only = [...single][0]
          viol.push(
            violation(
              "G3",
              fieldId,
              `별칭 '${value}'은 ${aliasCands.length}개 품목에 해당하나 규격 없이 ${only}로 확정`,
            ),
          )
        } else if (candSpecs.length === 1 && new Set(aliasCands.map((c) => c.spec)).size > 1) {
          viol.push(
            violation(
              "G3",
              fieldId,
              `별칭 '${value}'의 후보에 원문에 없는 규격 '${candSpecs[0]}'을 붙여 한 품목으로 좁힘`,
            ),
          )
        }
      }
    }

    if (item.humanFilled) {
      viol.push(violation("G5", fieldId, `사람 확정 칸에 AI 값 '${item.humanFilled}'`))
    }
  }

  if (extraction.humanFilled) {
    viol.push(violation("G5", null, "문서 수준 사람 확정 칸이 채워져 있음"))
  }

  return viol
}
